classes/submission/SubmissionDAO.inc.php

```diff
@@ -94,30 +95,85 @@ function newDataObject() {
 	/**
 	 * Internal function to return a Submission object from a row.
 	 * @param $row array
+	 * @param $submissionRevision int optional
 	 * @return Submission
 	 */
-	function _fromRow($row) {
+	function _fromRow($row, $submissionRevision = null) {
 		$submission = $this->newDataObject();
+		$submissionRevision = $submissionRevision ? $submissionRevision : $this->getLatestRevisionId($row['submission_id']);
 
 		$submission->setId($row['submission_id']);
+		$submission->setSubmissionRevision($submissionRevision);
 		$submission->setContextId($row['context_id']);
 		$submission->setLocale($row['locale']);
 		$submission->setStageId($row['stage_id']);
 		$submission->setStatus($row['status']);
 		$submission->setSubmissionProgress($row['submission_progress']);
 		$submission->setDateSubmitted($this->datetimeFromDB($row['date_submitted']));
 		$submission->setDateStatusModified($this->datetimeFromDB($row['date_status_modified']));
-		$submission->setDatePublished($this->datetimeFromDB($row['date_published']));
 		$submission->setLastModified($this->datetimeFromDB($row['last_modified']));
 		$submission->setLanguage($row['language']);
 		$submission->setCitations($row['citations']);
 
-		$this->getDataObjectSettings('submission_settings', 'submission_id', $submission->getId(), $submission);
+		$this->getDataObjectSettings('submission_settings', 'submission_id', $submission->getId(), $submission, $submissionRevision, 'submission_revision');
 
 		return $submission;
 	}
 
 	/**
+	 * Get all revisions for a submission
+	 * @param $submissionId int
+	 * @param $contextId int optional
+	 * $param $order string optional default: ASC
+	 * @return array
+	 */
+	function getSubmissionRevisionIds($submissionId, $contextId = null, $order = SORT_DIRECTION_ASC) {
+		$params = array((int) $submissionId);
+		if ($contextId) {
+			$params[] = (int) $contextId;
+		}
+
+		$sql = 'SELECT DISTINCT ss.submission_revision
+			FROM 	submission_settings ss
+				INNER JOIN submissions s ON ss.submission_id = s.submission_id
+			WHERE 	ss.submission_id = ?
+				'.($contextId ? ' AND s.context_id = ?' : '') .'
+			ORDER BY ss.submission_revision ' . $this->getDirectionMapping($order);
+
+		$result = $this->retrieve($sql, $params);
+		$submissionRevisions = array();
+
+		foreach($result as $submission) {
```

Comment:
S. https://github.com/pkp/pkp-lib/blob/master/lib/adodb/adodb.inc.php#L3499. I suppose that your DB supports arrays but maybe some other does not, so maybe to stay with the usual methods/calls?